aws-lc-rs/src/cmac.rs

```diff
@@ -0,0 +1,608 @@
+//! // back to us. We need to verify they didn't tamper with it.]
+//!
+//! cmac::verify(&key, msg.as_bytes(), tag.as_ref())?;
+//!
+//! # Ok::<(), aws_lc_rs::error::Unspecified>(())
+//! ```
+//!
+//! ## Using the one-shot API:
+//!
+//! ```
+//! use aws_lc_rs::rand::SecureRandom;
+//! use aws_lc_rs::{cmac, rand};
+//!
+//! let msg = "hello, world";
+//!
+//! // The sender generates a secure key value and signs the message with it.
+//! // Note that in a real protocol, a key agreement protocol would be used to
+//! // derive `key_value`.
+//! let rng = rand::SystemRandom::new();
+//! let key_value: [u8; 16] = rand::generate(&rng)?.expose();
+//!
+//! let s_key = cmac::Key::new(cmac::AES_128, key_value.as_ref());
+//! let tag = cmac::sign(&s_key, msg.as_bytes());
+//!
+//! // The receiver (somehow!) knows the key value, and uses it to verify the
+//! // integrity of the message.
+//! let v_key = cmac::Key::new(cmac::AES_128, key_value.as_ref());
+//! cmac::verify(&v_key, msg.as_bytes(), tag.as_ref())?;
+//!
+//! # Ok::<(), aws_lc_rs::error::Unspecified>(())
+//! ```
+//!
+//! ## Using the multi-part API:
+//! ```
+//! use aws_lc_rs::rand::SecureRandom;
+//! use aws_lc_rs::{cmac, rand};
+//!
+//! let parts = ["hello", ", ", "world"];
+//!
+//! // The sender generates a secure key value and signs the message with it.
+//! // Note that in a real protocol, a key agreement protocol would be used to
+//! // derive `key_value`.
+//! let rng = rand::SystemRandom::new();
+//! let key_value: [u8; 32] = rand::generate(&rng)?.expose();
+//!
+//! let s_key = cmac::Key::new(cmac::AES_256, key_value.as_ref());
+//! let mut s_ctx = cmac::Context::with_key(&s_key);
+//! for part in &parts {
+//!     s_ctx.update(part.as_bytes());
+//! }
+//! let tag = s_ctx.sign();
+//!
+//! // The receiver (somehow!) knows the key value, and uses it to verify the
+//! // integrity of the message.
+//! let v_key = cmac::Key::new(cmac::AES_256, key_value.as_ref());
+//! let mut msg = Vec::<u8>::new();
+//! for part in &parts {
+//!     msg.extend(part.as_bytes());
+//! }
+//! cmac::verify(&v_key, &msg.as_ref(), tag.as_ref())?;
+//!
+//! # Ok::<(), aws_lc_rs::error::Unspecified>(())
+//! ```
+//! [RFC 4493]: https://tools.ietf.org/html/rfc4493
+//! [NIST SP 800-38B]: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-38b.pdf
+
+use crate::aws_lc::{
+    CMAC_CTX_copy, CMAC_CTX_new, CMAC_Final, CMAC_Init,
+    CMAC_Update, EVP_aes_128_cbc, EVP_aes_192_cbc, EVP_aes_256_cbc, EVP_des_ede3_cbc, CMAC_CTX,
+};
+use crate::error::Unspecified;
+use crate::fips::indicator_check;
+use crate::ptr::LcPtr;
+use crate::{constant_time, rand};
+use core::mem::MaybeUninit;
+use core::ptr::null_mut;
+
+/// A CMAC algorithm.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct Algorithm {
+    name: &'static str,
+    key_len: usize,
+    tag_len: usize,
+    cipher_fn: unsafe extern "C" fn() -> *const crate::aws_lc::EVP_CIPHER,
```

Comment:
I'm not crazy about holding a function pointer here. I'd prefer the approach of using a (private) `AlgorithmId` type with a method that provides this value in the form of a `ConstPointer<'_, EVP_CIPHER>`, similar to what is done [here](https://github.com/aws/aws-lc-rs/blob/d4be093dee6572097f239bfabef144dd7f5a7930/aws-lc-rs/src/cipher.rs#L286-L306).

Reply:
I disagree. There’s an established pattern in our library of using an enum to represent “AlgorithmId”s, largely for this purpose.